File: src/3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */

export function lengthOfLongestSubstring(s) {
  const seen = new Map()
  let start = 0,
    maxLen = 0
  for (let i = 0; i < s.length; i++) {
    let val = s[i]
    if (seen.has(val)) {
      start = Math.max(start, seen.get(val) + 1)
    }
    maxLen = Math.max(maxLen, i - start + 1)
    seen.set(val, i)
  }
  return maxLen
}

File: src/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        map = {}
        maxLen = j = 0
        for i in range(n):
            c = s[i]
            if c in map:
                j = max(j, map[c] + 1)
            maxLen = max(maxLen, i - j + 1)
            map[s[i]] = i
        return maxLen